pkg/controller/nodeipam/ipam/multi_cidr_range_allocator.go

```diff
@@ -0,0 +1,1105 @@
+	eventBroadcaster.StartRecordingToSink(
+		&v1core.EventSinkImpl{
+			Interface: client.CoreV1().Events(""),
+		})
+
+	if client.CoreV1().RESTClient().GetRateLimiter() != nil {
+		ratelimiter.RegisterMetricAndTrackRateLimiterUsage("multi_cidr_range_allocator", client.CoreV1().RESTClient().GetRateLimiter())
+	}
+
+	ra := &multiCIDRRangeAllocator{
+		client:                  client,
+		nodeLister:              nodeInformer.Lister(),
+		nodesSynced:             nodeInformer.Informer().HasSynced,
+		clusterCIDRConfigLister: clusterCIDRConfigInformer.Lister(),
+		clusterCIDRConfigSynced: clusterCIDRConfigInformer.Informer().HasSynced,
+		nodeCIDRUpdateChannel:   make(chan multiCIDRNodeReservedCIDRs, cidrUpdateQueueSize),
+		recorder:                recorder,
+		nodesInProcessing:       sets.NewString(),
+		queue:                   workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "multi_cidr_range_allocator"),
+		lock:                    &sync.Mutex{},
+		cidrMap:                 make(map[string][]*cidrset.ClusterCIDR, 0),
+	}
+
+	// This is only set for testing purposes.
+	if len(allocatorParams.CIDRMap) > 0 {
+		ra.cidrMap = allocatorParams.CIDRMap
+	}
+
+	if cccList == nil {
+		cccList = &networkingv1alpha1.ClusterCIDRConfigList{}
+	}
+	createDefaultClusterCIDRConfig(cccList, allocatorParams)
+
+	// Regenerate the cidrMaps from the existing ClusterCIDRConfigs.
+	for _, clusterCIDRConfig := range cccList.Items {
+		klog.Infof("Creating ClusterCIDRConfig: %s", clusterCIDRConfig.Name)
+		// Create an event for invalid ClusterCIDRConfigs, do not crash on failures.
+		if err := ra.reconcileBootstrap(&clusterCIDRConfig); err != nil {
+			klog.Errorf("Error while creating ClusterCIDRConfig: %v", err)
+			ra.recorder.Event(&clusterCIDRConfig, "Warning", "InvalidClusterCIDRConfig", err.Error())
+		}
+	}
+
+	clusterCIDRConfigInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    cccutil.CreateClusterCIDRConfigHandler(ra.reconcileCreate),
+		DeleteFunc: cccutil.CreateClusterCIDRConfigHandler(ra.reconcileDelete),
+	})
+
+	if allocatorParams.ServiceCIDR != nil {
+		ra.filterOutServiceRange(allocatorParams.ServiceCIDR)
+	} else {
+		klog.V(0).Info("No Service CIDR provided. Skipping filtering out service addresses.")
+	}
+
+	if allocatorParams.SecondaryServiceCIDR != nil {
+		ra.filterOutServiceRange(allocatorParams.SecondaryServiceCIDR)
+	} else {
+		klog.V(0).Info("No Secondary Service CIDR provided. Skipping filtering out secondary service addresses.")
+	}
+
+	if nodeList != nil {
+		for _, node := range nodeList.Items {
+			if len(node.Spec.PodCIDRs) == 0 {
+				klog.V(4).Infof("Node %v has no CIDR, ignoring", node.Name)
+				continue
+			}
+			klog.V(0).Infof("Node %v has CIDR %s, occupying it in CIDR map", node.Name, node.Spec.PodCIDRs)
+			if err := ra.occupyCIDRs(&node); err != nil {
+				// This will happen if:
+				// 1. We find garbage in the podCIDRs field. Retrying is useless.
```

Comment:
We should think about how to handle this. 

Should a single Node having garbage in the PodCIDR field lock up the entire cluster?

At scale the answer is no. Can we figure out what the proper behavior is? 
This can be done in a follow on PR, but I don't want to drop handling this case.

Reply:
comment stands...

Reply:
The same issue is present in the existing range allocator implementation: https://github.com/kubernetes/kubernetes/blob/bc4c4930ff18dce320bd33f645ff4f3accc72018/pkg/controller/nodeipam/ipam/range_allocator.go#L127-L133

Not sure about the history here, why the decision was made to crashloop the range allocator, however is it ok for alpha for this to error out and then decide for both the range allocators and create a follow up post alpha

Reply:
:) from Tim

> It's not clear exactly what the FIX is, but it surely should not CRASH, even if it is a bad value.

https://github.com/kubernetes/kubernetes/issues/95948#issuecomment-742835516

Reply:
better log an error and continue, so we don't perpetuate the original problem

Reply:
you are already doing that in 
```

	// Regenerate the cidrMaps from the existing ClusterCIDRs.
	for _, clusterCIDR := range ccList.Items {
		klog.Infof("Regenerating existing ClusterCIDR: %v", clusterCIDR)
		// Create an event for invalid ClusterCIDRs, do not crash on failures.
		if err := ra.reconcileBootstrap(&clusterCIDR); err != nil {
			klog.Errorf("Error while regenerating existing ClusterCIDR: %v", err)
			ra.recorder.Event(&clusterCIDR, "Warning", "InvalidClusterCIDR encountered while regenerating ClusterCIDR during bootstrap.", err.Error())
		}
```